### PR TITLE
[Feature] Classic Taunt (Pre 2006) style of taunt

### PR DESCRIPTION
**Pull request by fryguy503**
RULE: ClassicTauntSystem

Completely seperate logic for each system. Could potentially be re-writte to be more inline but there are a good bit of minor differences.

### CODE REVIEW

File: zone/special_attacks.cpp [hunk 1]
```diff
@@ -2128,20 +2128,33 @@ void Client::DoClassAttacks(Mob *ca_target, uint16 skill, bool IsRiposte)
...
+	* something that was also above level 60."
+	* Also: "The chance to taunt an NPC higher level than yourself dropped off at double the rate
+	* if you were above level 60 than if you were below level 60 making it very hard to taunt creature
+	* higher level than yourself if you were above level 60."
+	* 
+	* See http://www.elitegamerslounge.com/home/soearchive/viewtopic.php?t=81156 */
 void Mob::Taunt(NPC *who, bool always_succeed, int chance_bonus, bool from_spell, int32 bonus_hate)
 {
-	if (!who || DivineAura() || (!from_spell && !CombatRange(who))) {
+	if (who == nullptr || DivineAura() || (!from_spell && !CombatRange(who)) || (IsNPC() && IsCharmed()) || !CombatRange(who)) {
```
Comment (Kinglykrab) [inline zone/special_attacks.cpp:2146]: `!who`.

**Review by Kinglykrab — APPROVED** [review]

File: zone/special_attacks.cpp [hunk 2]
```diff
@@ -2128,20 +2128,33 @@ void Client::DoClassAttacks(Mob *ca_target, uint16 skill, bool IsRiposte)
...
+	* something that was also above level 60."
+	* Also: "The chance to taunt an NPC higher level than yourself dropped off at double the rate
+	* if you were above level 60 than if you were below level 60 making it very hard to taunt creature
+	* higher level than yourself if you were above level 60."
+	* 
+	* See http://www.elitegamerslounge.com/home/soearchive/viewtopic.php?t=81156 */
 void Mob::Taunt(NPC *who, bool always_succeed, int chance_bonus, bool from_spell, int32 bonus_hate)
 {
-	if (!who || DivineAura() || (!from_spell && !CombatRange(who))) {
+	if (!who || DivineAura() || (!from_spell && !CombatRange(who)) || (IsNPC() && IsCharmed()) || !CombatRange(who)) {
```
Comment (Akkadius) [inline zone/special_attacks.cpp:2146]: What's the difference between the `!CombatRange` from `(!from_spell && !CombatRange(who))` and the catch all `!CombatRange(who)`

File: zone/special_attacks.cpp [hunk 3]
```diff
@@ -2153,104 +2166,133 @@ void Mob::Taunt(NPC *who, bool always_succeed, int chance_bonus, bool from_spell
...
+		// Taunt Chance Rule Bonus
+		taunt_chance += RuleI(Combat, TauntChanceBonus);
+	}
+
+	//success roll
+	success = zone->random.Roll(taunt_chance);
+
+	// Log result
+	LogHate(
+		"Taunter mob {} target npc {} taunt_chance [{}] success [{}] hate_top [{}]",
```
Comment (Akkadius) [inline zone/special_attacks.cpp:2240]: Enclose all `{}` within brackets please `[{}]`

File: zone/special_attacks.cpp [hunk 4]
```diff
@@ -2153,104 +2166,133 @@ void Mob::Taunt(NPC *who, bool always_succeed, int chance_bonus, bool from_spell
...
+
+		MessageString(Chat::Skills, TAUNT_SUCCESS, who->GetCleanName());
 	} else {
-		MessageString(Chat::SpellFailure, FAILED_TAUNT);
+		MessageString(Chat::Skills, FAILED_TAUNT);
 	}
 
-	TryCastOnSkillUse(who, EQ::skills::SkillTaunt);
+	// Modern Abilities
+	if (RuleB(Combat, ClassicTauntSystem)) {
```
Comment (Akkadius) [inline zone/special_attacks.cpp:2286]: If this section is supposed to be for modern abilities, wouldn't this be `if (!RuleB(Combat, ClassicTauntSystem)) {` ?

Comment (fryguy503) [reply zone/special_attacks.cpp:2286]: Good Catch, thank you.

Comment (fryguy503) [reply zone/special_attacks.cpp:2286]: Wonder if I should rule this off separately or remove the rule all together. This is less for the taunting itself and more for additional features.

Comment (Akkadius) [reply zone/special_attacks.cpp]: Generally - the more you can avoid deeply nested if/else chains the better. It quickly makes code harder to read and reason about. For this change specifically I would suggest the following

```cpp
	// classic taunt did not utilize modern abilities such as skill procs
	if (RuleB(Combat, ClassicTauntSystem)) {
		return;
	}
```

Comment (Akkadius) [reply zone/special_attacks.cpp]: image

Comment (fryguy503) [reply zone/special_attacks.cpp]: I agree with cleaning up nested ifs for sure.

What I am referencing is the `TrySkillProc` is less directly tied to taunt or classic taunt and is a separate system and I am thinking that it shouldn't be ruled out as part of the classic taunt system. If someone wants to have SkillProcs from Taunt and still use the classic taunt system they should be able to.

**Pull request merged**